lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,659 @@
+    const auto FindCurrentNode = [&CurrentNode](const auto &Pair) {
+      return Pair.Neighbor == CurrentNode;
+    };
+    auto EdgeIt = llvm::find_if(ChildrenEdgesRange, FindCurrentNode);
+    revng_assert(EdgeIt != ChildrenEdgesRange.end());
+    EdgeIt->Label->setType(EdgeType::TreeEdge);
+  }
+
+  // All the remaining edges that have not been marked as `TreeEdge`s, are the
+  // `BackEdge`s. Mind that we cannot perform this assignment during the
+  // previous visit itself, because we first need to visit all the edges to mark
+  // the `TreeEdge`s
+  for (BlockNode *CurrentNode :
+       llvm::depth_first(llvm::Undirected(Graph.getEntryNode()))) {
+    for (auto &[Neighbor, Label] :
+         llvm::children_edges<llvm::Undirected<BlockNode *>>(CurrentNode)) {
+      if (Label->type() == EdgeType::Invalid) {
+        Label->setType(EdgeType::BackEdge);
+      }
+    }
+  }
+}
+
+/// Helper function to materialize the capping backedges
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::insertCappingBackedges(BlockGraph
+                                                                    &Graph) {
+  // Highest, i.e., closer to the spanning tree root, reached DFSNum for each
+  // `BB`
+  std::map<BlockNode *, size_t> HighestReachedDFSNum;
+
+  // Keep a mapping between the `DFSNum` and the corresponding node
+  std::map<size_t, BlockNode *> DFSNumNodeMap;
+
+  llvm::SmallVector<BlockNode *> PostOrder;
+  for (BlockNode *BB :
+       llvm::post_order(llvm::Undirected<BlockNode *>(Graph.getEntryNode()))) {
+    PostOrder.push_back(BB);
+  }
+
+  for (BlockNode *BB : PostOrder) {
+
+    // 1) Compute the highest reached node from a backedge starting from the
+    //    current `Block`
+    size_t HI0 = SizeTMaxValue;
+
+    // We are only interested in looking at backedges
+    for (const auto &[BackwardEdge, _, __] : getBackwardEdges(BB)) {
+      BlockNode *Target = std::get<1>(BackwardEdge);
+      HI0 = std::min(HI0, Target->getDFSNum());
+
+      // We also populate the mapping between the `DFSNum` and the node
+      DFSNumNodeMap[Target->getDFSNum()] = Target;
+    }
+
+    // 2) Compute the highest reached node for any child of the current `Block`
+    size_t HI1 = SizeTMaxValue;
+    for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
+      BlockNode *Target = std::get<1>(ForwardEdge);
+      HI1 = std::min(HI1, HighestReachedDFSNum.at(Target));
+    }
+
+    // 3) Assign the highest reached node for the current `Block`
+    HighestReachedDFSNum[BB] = std::min(HI0, HI1);
+
+    // 4) Compute the child which has the reached DFSNum computed at step 2)
+    BlockNode *HI1Child = nullptr;
+    if (HI1 != SizeTMaxValue) {
+      for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
+        BlockNode *Target = std::get<1>(ForwardEdge);
+        if (HI1 == HighestReachedDFSNum.at(Target)) {
+          HI1Child = Target;
+          break;
+        }
+      }
+    }
```

Comment:
I actually think that you can merge 2 and 4 in a single loop, very similar to what I proposed for point 5.

And do point 3 after old point 2+4(merged) and before point 5.